sys/ecc/golay2412.c

```diff
@@ -0,0 +1,433 @@
+/*
+ * Copyright (c) 2007 - 2015 Joseph Gaeddert
+ * Copyright (c) 2018        HAW Hamburg
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+/**
+ * @ingroup     sys_ecc
+ * @{
+ *
+ * @file
+ * @brief       Golay(24,12) half-rate forward error-correction code
+ *
+ * Code referring to Liquid-DSP
+ * https://github.com/jgaeddert/liquid-dsp/blob/master/src/fec/src/fec_golay2412.c
+ *
+ * @author      Joseph Gaeddert
+ * @author      Peter Kietzmann <peter.kietzmann@haw.hamburg.de>
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <errno.h>
+
+#include "bitarithm.h"
+#include "ecc/golay2412.h"
+
+#define ENABLE_DEBUG        (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+#define DEBUG_FEC_GOLAY2412 (1)
+#else
+#define DEBUG_FEC_GOLAY2412 (0)
+#endif
+
+/* generator matrix transposed [24 x 12] */
+static const uint32_t golay2412_Gt[24] = {
+    0x08ed, 0x01db, 0x03b5, 0x0769, 0x0ed1, 0x0da3, 0x0b47, 0x068f,
+    0x0d1d, 0x0a3b, 0x0477, 0x0ffe, 0x0800, 0x0400, 0x0200, 0x0100,
+    0x0080, 0x0040, 0x0020, 0x0010, 0x0008, 0x0004, 0x0002, 0x0001};
+
+/* P matrix [12 x 12] */
+static const uint32_t golay2412_P[12] = {
+    0x08ed, 0x01db, 0x03b5, 0x0769,
+    0x0ed1, 0x0da3, 0x0b47, 0x068f,
+    0x0d1d, 0x0a3b, 0x0477, 0x0ffe};
+
+/* parity check matrix [12 x 24] */
+static const uint32_t golay2412_H[12] = {
+    0x008008ed, 0x004001db, 0x002003b5, 0x00100769,
+    0x00080ed1, 0x00040da3, 0x00020b47, 0x0001068f,
+    0x00008d1d, 0x00004a3b, 0x00002477, 0x00001ffe};
+
+#if DEBUG_FEC_GOLAY2412
+/* print string of bits to standard output */
+static inline void liquid_print_bitstring(uint32_t _x,
+                            uint32_t _n)
+{
+    uint32_t i;
+    for (i=0; i<_n; i++)
+        printf("%" PRIu32, (_x >> (_n-i-1)) & 1);
+}
+#endif
+
+static inline uint32_t block_get_enc_msg_len(uint32_t _dec_msg_len,
```

Comment:
This looks rather long for an inline function. Are you sure you intentended it to be inline?